File: core/src/main/scala/org/trustedanalytics/sparktk/models/kmeans/KMeans.scala
package org.trustedanalytics.sparktk.models.kmeans

import org.apache.spark.SparkContext
import org.apache.spark.mllib.clustering.{ KMeans => SparkKMeans, KMeansModel => SparkKMeansModel }
import org.apache.spark.mllib.linalg.{ Vector => MllibVector }
import org.apache.spark.sql.{ SQLContext, Row }
import org.trustedanalytics.sparktk.frame.internal.RowWrapper
import org.trustedanalytics.sparktk.frame.internal.rdd.{ VectorUtils, FrameRdd, RowWrapperFunctions }
import org.trustedanalytics.sparktk.frame._

import scala.language.implicitConversions

object KMeans {

  /**
   * @param frame The frame containing the data to train on
   * @param columns The columns to train on
   * @param k Desired number of clusters
   * @param scalings Optional scaling values, each is multiplied by the corresponding value in the observation column
   * @param maxIterations Number of iterations for which the algorithm should run
   * @param epsilon Distance threshold within which we consider k-means to have converged. Default is 1e-4. If all
   *                centers move less than this Euclidean distance, we stop iterating one run
   * @param initializationMode The initialization technique for the algorithm.  It could be either "random" to
   *                           choose random points as initial clusters, or "k-means||" to use a parallel variant
   *                           of k-means++.  Default is "k-means||"
   */
  def train(frame: Frame,
            columns: Seq[String],
            k: Int = 2,
            scalings: Option[Seq[Double]] = None,
            maxIterations: Int = 20,
            epsilon: Double = 1e-4,
            initializationMode: String = "k-means||",
            seed: Option[Long] = None): KMeansModel = {

    require(columns != null && columns.nonEmpty, "columns must not be null nor empty")
    require(scalings != null, "scalings must not be null")
    if (scalings.isDefined) {
      require(columns.length == scalings.get.length, "Length of columns and scalings needs to be the same")
    }
    require(k > 0, "k must be at least 1")
    require(maxIterations > 0, "maxIterations must be a positive value")
    require(epsilon > 0.0, "epsilon must be a positive value")
    require(initializationMode == "random" || initializationMode == "k-means||", "initialization mode must be 'random' or 'k-means||'")

    val sparkKMeans = new SparkKMeans()
    sparkKMeans.setK(k)
    sparkKMeans.setMaxIterations(maxIterations)
    sparkKMeans.setInitializationMode(initializationMode)
    sparkKMeans.setEpsilon(epsilon)
    if (seed.isDefined) {
      sparkKMeans.setSeed(seed.get)
    }

    val trainFrameRdd = new FrameRdd(frame.schema, frame.rdd)
    trainFrameRdd.cache()
    val vectorRDD = scalings match {
      case Some(weights) => trainFrameRdd.toDenseVectorRddWithWeights(columns, weights)
      case None => trainFrameRdd.toDenseVectorRdd(columns)
    }
    val model = sparkKMeans.run(vectorRDD)
    val centroids: Array[Array[Double]] = model.clusterCenters.map(_.toArray) // Make centroids easy for Python
    trainFrameRdd.unpersist()

    //Writing the kmeansModel as JSON
    //val jsonModel = new KMeansData(kmeansModel, arguments.observationColumns, arguments.columnScalings)
    //val model: Model = arguments.model
    //model.data = jsonModel.toJson.asJsObject

    KMeansModel(columns, k, scalings, maxIterations, epsilon, initializationMode, centroids, model)
  }

}

/**
 * @param k Desired number of clusters
 * @param scalings Optional scaling values, each is multiplied by the corresponding value in the observation column
 * @param maxIterations Number of iteration for which the algorithm should run
 * @param epsilon Distance threshold within which we consider k-means to have converged. Default is 1e-4. If all
 *                centers move less than this Euclidean distance, we stop iterating one run
 * @param initializationMode The initialization technique for the algorithm.  It could be either "random" to
 *                           choose random points as initial clusters, or "k-means||" to use a parallel variant
 *                           of k-means++.  Default is "k-means||"
 * @param centroids An array of length k containing the centroid of each cluster
 */
case class KMeansModel private[kmeans] (columns: Seq[String],
                                        k: Int,
                                        scalings: Option[Seq[Double]],
                                        maxIterations: Int,
                                        epsilon: Double,
                                        initializationMode: String,
                                        centroids: Array[Array[Double]],
                                        private val model: SparkKMeansModel) extends Serializable {

  implicit def rowWrapperToRowWrapperFunctions(rowWrapper: RowWrapper): RowWrapperFunctions = {
    new RowWrapperFunctions(rowWrapper)
  }

  /**
   * Helper method that provides a function to create an appropriate MllibVector for a RowWrapper, taking into account optional weights
   * @param observationColumns
   * @return
   */
  private def getDenseVectorMaker(observationColumns: Seq[String]): RowWrapper => MllibVector = {

    def getDenseVector(columnNames: Seq[String])(row: RowWrapper): MllibVector = {
      row.toDenseVector(columnNames)
    }

    def getWeightedDenseVector(columnNames: Seq[String], columnWeights: Array[Double])(row: RowWrapper): MllibVector = {
      row.toWeightedDenseVector(columnNames, columnWeights)
    }

    scalings match {
      case None => getDenseVector(observationColumns)
      case Some(weights) =>
        require(weights.length == observationColumns.length)
        getWeightedDenseVector(observationColumns, weights.toArray)
    }
  }

  /**
   * Computes the number of elements belonging to each cluster given the trained model and names of the frame's columns storing the observations
   * @param frame A frame containing observations
   * @param observationColumns The columns of frame storing the observations (uses column names from train by default)
   * @return An array of length k of the cluster sizes
   */
  def computeClusterSizes(frame: Frame, observationColumns: Option[Seq[String]] = None): Array[Int] = {
    require(frame != null, "frame is required")
    if (observationColumns.isDefined) {
      require(columns.length == observationColumns.get.length, "Number of columns for train and predict should be same")
    }

    val vectorMaker = getDenseVectorMaker(observationColumns.getOrElse(columns))
    val frameRdd = new FrameRdd(frame.schema, frame.rdd)
    val predictRDD = frameRdd.mapRows(row => {
      val point = vectorMaker(row)
      model.predict(point)
    })
    val clusterSizes = predictRDD.map(row => (row.toString, 1)).reduceByKey(_ + _).collect().map { case (k, v) => v }
    clusterSizes
  }

  /**
   * Computes the 'within cluster sum of squared errors' for the given frame
   * @param frame A frame containing observations
   * @param observationColumns The columns of frame storing the observations (uses column names from train by default)
   * @return wsse
   */
  def computeWsse(frame: Frame, observationColumns: Option[Vector[String]] = None): Double = {
    require(frame != null, "frame is required")
    if (observationColumns.isDefined) {
      require(columns.length == observationColumns.get.length, "Number of columns for train and predict should be same")
    }

    val frameRdd = new FrameRdd(frame.schema, frame.rdd)
    val vectorRdd = frameRdd.toDenseVectorRdd(observationColumns.getOrElse(columns), scalings)
    model.computeCost(vectorRdd)
  }

  def addDistanceColumns(frame: Frame, observationColumns: Option[Vector[String]] = None): Unit = {
    require(frame != null, "frame is required")
    if (observationColumns.isDefined) {
      require(columns.length == observationColumns.get.length, "Number of columns for train and predict should be same")
    }

    val vectorMaker = getDenseVectorMaker(observationColumns.getOrElse(columns))
    val distanceMapper: RowWrapper => Row = row => {
      val point = vectorMaker(row)
      val clusterCenters = model.clusterCenters
      Row.fromSeq(for (i <- clusterCenters.indices) yield {
        val distance: Double = VectorUtils.toMahoutVector(point).getDistanceSquared(VectorUtils.toMahoutVector(clusterCenters(i)))
        distance
      })
    }

    val newColumns = (for (i <- model.clusterCenters.indices) yield Column("distance" + i.toString, DataTypes.float64)).toSeq
    frame.addColumns(distanceMapper, newColumns)
  }

  /**
   * Adds a column to the frame which indicates the predicted cluster for each observation
   * @param frame - frame to add predictions to
   * @param observationColumns Column(s) containing the observations whose clusters are to be predicted.
   *                           Default is to predict the clusters over columns the KMeans model was trained on.
   *                           The columns are scaled using the same values used when training the model
   */
  def predict(frame: Frame, observationColumns: Option[Vector[String]] = None): Unit = {
    require(frame != null, "frame is required")
    if (observationColumns.isDefined) {
      require(columns.length == observationColumns.get.length, "Number of columns for train and predict should be same")
    }

    val vectorMaker = getDenseVectorMaker(observationColumns.getOrElse(columns))
    val predictMapper: RowWrapper => Row = row => {
      val point = vectorMaker(row)
      val prediction = model.predict(point)
      Row.apply(prediction)
    }

    frame.addColumns(predictMapper, Seq(Column("cluster", DataTypes.int32)))
  }

}

//object KMeansModel {
//
//  def save(sc: SparkContext, model: KMeansModel, path: String): Unit = {
//    val sqlContext = SQLContext.getOrCreate(sc)
//    import sqlContext.implicits._
//    val metadata = compact(render(
//      ("class" -> thisClassName) ~ ("version" -> thisFormatVersion) ~ ("k" -> model.k)))
//    sc.parallelize(Seq(metadata), 1).saveAsTextFile(Loader.metadataPath(path))
//    val dataRDD = sc.parallelize(model.clusterCenters.zipWithIndex).map { case (point, id) =>
//      Cluster(id, point)
//    }.toDF()
//    dataRDD.write.parquet(Loader.dataPath(path))
//  }
//
//  def load(modelBytes: Array[Byte]): KMeansModel = {
//    KMeansModel(Vector[String]("a", "b"), k=3, scalings=None, maxIterations = 20, epsilon = 1.1, initializationMode = "k-means||", centroids=null, model=null)
//  }
//}
//
//case class KMeansWellModel(columns: List[String], k: Int) {
//  def className = productPrefix // gets the "name" of this case class
//}
//
//object KMeansWellModel {
//  private lazy val dummyModel = KMeansWellModel(Nil, 0)
//  def className = dummyModel.className
//}

File: python/sparktk/models/clustering/kmeans.py
from sparktk.loggers import log_load; log_load(__name__); del log_load

from sparktk.propobj import PropertiesObject
from sparktk import TkContext

__all__ = ["train", "load", "KMeansModel"]

def train(frame, columns, k=2, scalings=None, max_iter=20, epsilon=1e-4, seed=None, init_mode="k-means||"):
    """
    Creates a KMeansModel by training on the given frame

    :param frame: (Frame) frame of training data
    :param columns: (List[str]) names of columns containing the observations for training
    :param k: (Optional (int)) number of clusters
    :param scalings: (Optional(List[float])) column scalings for each of the observation columns.  The scaling value is multiplied by
     the corresponding value in the observation column
    :param max_iter: (Optional(int)) number of iterations for which the algorithm should run
    :param epsilon: (Optional(float)) distance threshold within which we consider k-means to have converged. Default is 1e-4.
     If all centers move less than this Euclidean distance, we stop iterating one run
    :param seed: Optional(long) seed for randomness
    :param init_mode: (Optional(str)) the initialization technique for the algorithm.   It can be either "random" to choose
     random points as initial clusters or "k-means||" to use a parallel variant of k-means++. Default is "k-means||
    :return: (KMeansModel) trained KMeans model

    """
    tc = frame._tc
    _scala_obj = get_scala_obj(tc)
    if isinstance(columns, basestring):
        columns = [columns]
    scala_columns = tc.jutils.convert.to_scala_vector_string(columns)
    if scalings:
        scala_scalings = tc.jutils.convert.to_scala_vector_double(scalings)
        scala_scalings = tc.jutils.convert.to_scala_option(scala_scalings)
    else:
        scala_scalings = tc.jutils.convert.to_scala_option(None)

    seed = seed if seed is None else long(seed)
    scala_seed = tc.jutils.convert.to_scala_option(seed)
    scala_model = _scala_obj.train(frame._scala, scala_columns, k, scala_scalings, max_iter, epsilon, init_mode, scala_seed)
    return KMeansModel(tc, scala_model)


def load(path, tc=TkContext.implicit):
    """load KMeansModel from given path"""
    TkContext.validate(tc)
    return tc.load(path, KMeansModel)


def get_scala_obj(tc):
    """Gets reference to the scala object"""
    return tc.sc._jvm.org.trustedanalytics.sparktk.models.clustering.kmeans.KMeansModel


class KMeansModel(PropertiesObject):
    """
    A trained KMeans model

    Example
    -------

        >>> frame = tc.frame.create([[2, "ab"],
        ...                          [1,"cd"],
        ...                          [7,"ef"],
        ...                          [1,"gh"],
        ...                          [9,"ij"],
        ...                          [2,"kl"],
        ...                          [0,"mn"],
        ...                          [6,"op"],
        ...                          [5,"qr"]],
        ...                         [("data", float), ("name", str)])

        >>> model = tc.models.clustering.kmeans.train(frame, ["data"], 3, seed=5)

        >>> model.k
        3

        >>> sizes = model.compute_sizes(frame)

        >>> sizes
        [2, 2, 5]

        >>> wsse = model.compute_wsse(frame)

        >>> wsse
        5.3

        >>> model.predict(frame)

        >>> frame.inspect()
        [#]  data  name  cluster
        ========================
        [0]   2.0  ab          1
        [1]   1.0  cd          1
        [2]   7.0  ef          0
        [3]   1.0  gh          1
        [4]   9.0  ij          0
        [5]   2.0  kl          1
        [6]   0.0  mn          1
        [7]   6.0  op          2
        [8]   5.0  qr          2


        >>> model.add_distance_columns(frame)

        >>> frame.inspect()
        [#]  data  name  cluster  distance0  distance1  distance2
        =========================================================
        [0]   2.0  ab          1       36.0       0.64      12.25
        [1]   1.0  cd          1       49.0       0.04      20.25
        [2]   7.0  ef          0        1.0      33.64       2.25
        [3]   1.0  gh          1       49.0       0.04      20.25
        [4]   9.0  ij          0        1.0      60.84      12.25
        [5]   2.0  kl          1       36.0       0.64      12.25
        [6]   0.0  mn          1       64.0       1.44      30.25
        [7]   6.0  op          2        4.0      23.04       0.25
        [8]   5.0  qr          2        9.0      14.44       0.25

        >>> model.columns
        [u'data']

        >>> model.scalings  # None


        >>> centroids = model.centroids

        >>> model.save("sandbox/kmeans1")

        >>> restored = tc.load("sandbox/kmeans1")

        >>> restored.centroids == centroids
        True

        >>> restored_sizes = restored.compute_sizes(frame)

        >>> restored_sizes == sizes
        True

    <hide>
    >>> restored2 = tc.models.clustering.kmeans.load("sandbox/kmeans1")

    >>> restored.centroids == centroids
    True

    >>> restored.predict(frame)
    [#]  data  name  cluster  cluster_0
    ===================================
    [0]   2.0  ab          1          1
    [1]   1.0  cd          1          1
    [2]   7.0  ef          0          0
    [3]   1.0  gh          1          1
    [4]   9.0  ij          0          0
    [5]   2.0  kl          1          1
    [6]   0.0  mn          1          1
    [7]   6.0  op          2          2
    [8]   5.0  qr          2          2


    </hide>
    """

    def __init__(self, tc, scala_model):
        self._tc = tc
        tc.jutils.validate_is_jvm_instance_of(scala_model, get_scala_obj(tc))
        self._scala = scala_model

    @staticmethod
    def _from_scala(tc, scala_model):
        return KMeansModel(tc, scala_model)

    @property
    def columns(self):
        return list(self._tc.jutils.convert.from_scala_seq(self._scala.columns()))

    @property
    def scalings(self):
        s = self._tc.jutils.convert.from_scala_option(self._scala.scalings())
        if s:
            return list(self._tc.jutils.convert.from_scala_seq(s))
        return None

    @property
    def k(self):
        return self._scala.k()

    @property
    def max_iterations(self):
        return self._scala.maxIterations()

    @property
    def initialization_mode(self):
        return self._scala.initializationMode()

    @property
    def centroids(self):
        return [list(item) for item in list(self._scala.centroidsAsArrays())]

    def compute_sizes(self, frame, columns=None):
        c = self.__columns_to_option(columns)
        return [int(n) for n in self._scala.computeClusterSizes(frame._scala, c)]

    def compute_wsse(self, frame, columns=None):
        c = self.__columns_to_option(columns)
        return self._scala.computeWsse(frame._scala, c)

    def predict(self, frame, columns=None):
        c = self.__columns_to_option(columns)
        self._scala.predict(frame._scala, c)

    def add_distance_columns(self, frame, columns=None):
        c = self.__columns_to_option(columns)
        self._scala.addDistanceColumns(frame._scala, c)

    def __columns_to_option(self, columns):
        if isinstance(columns, basestring):
            columns = [columns]
        if columns is not None:
            columns = self._tc.jutils.convert.to_scala_vector_string(columns)
        return self._tc.jutils.convert.to_scala_option(columns)

    def save(self, path):
        if isinstance(path, basestring):
            self._scala.save(self._tc._scala_sc, path)

    def export_to_mar(self, path):
        """ export the trained model to MAR format for Scoring Engine """
        if isinstance(path, basestring):
            return self._scala.exportToMar(self._tc._scala_sc, path)

del PropertiesObject
